org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeBinding.java

```diff
@@ -92,21 +92,32 @@ public abstract class JavacTypeBinding implements ITypeBinding {
 	public final TypeSymbol typeSymbol;
 	private final Types types;
 	private final Type type;
-	private boolean isDeclaration;
+	private boolean specificInstance;
 	private boolean recovered = false;
 
-	public JavacTypeBinding(final Type type, final TypeSymbol typeSymbol, boolean isDeclaration, JavacBindingResolver resolver) {
+	public JavacTypeBinding(final Type type, final TypeSymbol typeSymbol, boolean specificInstance, JavacBindingResolver resolver) {
 		if (type instanceof PackageType) {
 			throw new IllegalArgumentException("Use JavacPackageBinding");
 		}
 		this.type = type;
-		this.isDeclaration = isDeclaration;
+		this.specificInstance = specificInstance;
 		this.resolver = resolver;
 		this.types = Types.instance(this.resolver.context);
 		// TODO: consider getting rid of typeSymbol in constructor and always derive it from type
-		this.typeSymbol = typeSymbol.kind == Kind.ERR ? this.type.tsym : typeSymbol;
+		//this.typeSymbol = typeSymbol.kind == Kind.ERR ? this.type.tsym : typeSymbol;
+		this.typeSymbol = typeSymbol;
 	}
 
+	/*
+	 * Some times, we can determine whether a binding is related to a specific instance
+	 * at creation time. OTher times, like when we traverse the owners of a symbol up the chain,
+	 * we cannot. In this case we default to assuming it is related to a specific instance 
+	 * until proven otherwise later on. 
+	 */
+	public void clearSpecificInstance() {
+		this.specificInstance = false;
```

Comment:
This approach cannot work well as long as we use the key to cache:
Take for example `class E<T> { E<T> instance; }`, 1st is generic (specificInstance == false`), 2nd is parametereized (`specificInstance == true`).
If you ask the type binding for each of them, you should get different bindings with different keys: `Ltest1/E<TT;>;` and `Ltest1/E<Ltest1/E;:T;>;` for the 2nd one.
As the value of `specificInstance` affects how key is generated and thus whether we must cache a new instance or not, it must not be modified during the lifecycle of the object.